server/request.go

```diff
@@ -392,44 +400,154 @@ func (r *RpcRequest) _writeRpcResponse(res *JsonRpcResponse) {
+	if txStatus == types.TxStatusFailed {
+		return true
+	} else if txStatus == types.TxStatusUnknown && time.Since(timeSent).Minutes() >= 5 {
+		return true
+	} else {
+		// If tx is still pending, or included then don't send it again
+		return false
+	}
+}
+
 // Send tx to relay and finish request (write response)
 func (r *RpcRequest) sendTxToRelay() {
-	// Check if tx was already forwarded and should be blocked now
 	txHash := strings.ToLower(r.tx.Hash().Hex())
-	if !ShouldSendTxToRelay(txHash) {
+
+	// Check if tx was already forwarded and should be blocked now
+	if !r.shouldSendTxToRelay(txHash) {
 		r.log("[sendTxToRelay] shouldn't send %s", txHash)
 		r.writeRpcResult(txHash)
 		return
 	}
 
 	r.log("[sendTxToRelay] sending %s ...", txHash)
 
-	delete(State.txStatus, txHash)           // remove any previous tx status
-	State.txForwardedToRelay[txHash] = Now() // remember tx was forwarded to relay
+	// mark tx as sent to relay
+	err := RState.SetTxSentToRelay(txHash)
+	if err != nil {
+		r.logError("[sendTxToRelay] redis:SetTxSentToRelay failed: %v", err)
+	}
 
-	// for cancellation, remember that this tx was sent to relay
-	State.userTxWithNonceSentToRelay[fmt.Sprintf("%s_%d", strings.ToLower(r.txFrom), r.tx.Nonce())] = NewBoolWithTime(true)
+	// remember that this tx based on from+nonce (for cancel-tx)
+	err = RState.SetTxHashForSenderAndNonce(r.txFrom, r.tx.Nonce(), txHash)
+	if err != nil {
+		r.logError("[sendTxToRelay] redis:SetTxHashForSenderAndNonce failed: %v", err)
+	}
+
+	// err = RState.SetLastPrivTxHashOfAccount(r.txFrom, txHash)
+	// if err != nil {
+	// 	r.logError("[sendTxToRelay] redis:SetLastTxHashOfAccount failed: %v", err)
+	// }
 
 	if DebugDontSendTx {
 		r.log("faked sending tx to relay, did nothing")
-		r.writeRpcResult(r.tx.Hash().Hex())
+		r.writeRpcResult(txHash)
 		return
 	}
 
-	param := make(map[string]string)
-	param["tx"] = r.rawTxHex
-	jsonRpcReq := NewJsonRpcRequest1(1, "eth_sendPrivateTransaction", param)
-	backendResp, respBytes, err := SendRpcWithSignatureAndParseResponse(r.relayUrl, r.relaySigningKey, jsonRpcReq)
+	sendPrivTxArgs := flashbotsrpc.FlashbotsSendPrivateTransactionRequest{Tx: r.rawTxHex}
+	_, err = FlashbotsRPC.FlashbotsSendPrivateTransaction(r.relaySigningKey, sendPrivTxArgs)
 	if err != nil {
-		r.logError("[sendTxToRelay] relay call failed for %s: %s - data: %s", txHash, err, *respBytes)
+		if errors.Is(err, flashbotsrpc.ErrRelayErrorResponse) {
+			r.log("[sendTxToRelay] relay error response: %v - rawTx: %s", err, r.rawTxHex)
+			r.writeRpcError(err.Error())
+		} else {
+			r.logError("[sendTxToRelay] relay call failed: %v - rawTx: %s", err, r.rawTxHex)
+			r.writeHeaderStatus(http.StatusInternalServerError)
+		}
+		return
+	}
+
+	r.writeRpcResult(txHash)
+	r.log("[sendTxToRelay] sent %s", txHash)
+}
+
+// Sends cancel-tx to relay as cancelPrivateTransaction, if initial tx was sent there too.
+func (r *RpcRequest) handleCancelTx() (requestCompleted bool) {
```

Comment:
Good point. I've tried doing this just now (splitting into check and send function), but the issue I've encountered is that both need to call Redis `GetTxHashForSenderAndNonce`, to get the tx hash for sender+nonce, which doubles all the error handling around Redis, which I'd prefer to avoid in this case.

Reply:
Sure, if it's not easily splitable and you tried then I don't have any more inputs